test/pytest/test_bridge.py

```diff
@@ -630,3 +635,282 @@ async def test_large_upload(event_loop, transport, tmp_path):
+        self.state: JsonObject = {}
+
+    @classmethod
+    async def open(
+            cls,
+            transport: MockTransport,
+            path: 'str | Path',
+            attrs: Sequence[str] = ('type', 'target'),
+            fnmatch: str = '*.txt',
+            **kwargs: JsonValue
+    ) -> 'FsInfoClient':
+        channel = await transport.check_open('fsinfo', path=str(path), attrs=attrs,
+                                             fnmatch=fnmatch, reply_keys=None, **kwargs)
+        return cls(transport, channel)
+
+    async def next_state(self) -> JsonObject:
+        while True:
+            patch = await self.transport.next_msg(self.channel)
+            self.state = json_merge_patch(self.state, patch)
+            if not get_bool(self.state, 'partial', None):
+                break
+        return self.state
+
+    async def wait(self) -> JsonObject:
+        state = await self.next_state()
+        await self.transport.assert_msg('', command='done', channel=self.channel)
+        await self.transport.assert_msg('', command='close', channel=self.channel)
+        return state
+
+    async def close(self):
+        await self.transport.check_close(self.channel)
+
+
+def fsinfo_err(err: int) -> JsonObject:
+    problems = {
+        errno.ENOENT: 'not-found',
+        errno.EPERM: 'access-denied',
+        errno.EACCES: 'access-denied',
+        errno.ENOTDIR: 'not-directory'
+    }
+    return {
+        'error': {
+            'problem': problems.get(err, 'internal-error'),
+            'message': os.strerror(err),
+            'errno': errno.errorcode[err],
+        }
+    }
+
+
+@pytest.fixture
+def fsinfo_test_cases(tmp_path: Path) -> 'dict[Path, JsonObject]':
+    # a normal directory
+    normal_dir = tmp_path / 'dir'
+    normal_dir.mkdir()
+    (normal_dir / 'dir-file.txt').write_text('dir file')
+    (normal_dir / 'dir-file.xtx').write_text('do not read this')
+    dir_entries = {'dir-file.txt': {"type": "reg"}}
+
+    # a directory without +x (search)
+    no_x_dir = tmp_path / 'no-x-dir'
+    no_x_dir.mkdir()
+    (no_x_dir / 'no-x-dir-file.txt').write_text('file')
+    no_x_dir.chmod(0o644)
+
+    # a directory without +r (read)
+    no_r_dir = tmp_path / 'no-r-dir'
+    no_r_dir.mkdir()
+    (no_r_dir / 'no-r-dir-file.txt').write_text('file')
+    no_r_dir.chmod(0o311)
+
+    # a normal file
+    file = tmp_path / 'file'
+    file.write_text('normal file')
+
+    # a non-readable file
+    no_r_file = tmp_path / 'no-r-file'
+    no_r_file.write_text('inaccessible file')
+    no_r_file.chmod(0)
+
+    # a device
+    dev = tmp_path / 'dev'
+    dev.symlink_to('/dev/null')
+
+    # a dangling symlink
+    dangling = tmp_path / 'dangling'
+    dangling.symlink_to('does-not-exist')
+
+    # a symlink pointing to itself
+    loopy = tmp_path / 'loopy'
+    loopy.symlink_to(loopy)
+
+    expected_state: 'dict[Path, JsonObject]' = {
+        normal_dir: {"info": {"type": "dir", "entries": dir_entries}},
```

Comment:
Would you mind expanding `dir_entries` here? It's easier to read, and avoids the definition so far away from this. It looks rather odd up there, as it's not part of the test setup, but the expected result.